File: experiments/language-spike/haskell/src/Kernel.hs
module Kernel (run, decodeFailure) where

import Control.Monad (foldM)
import qualified Data.ByteString as BS
import qualified Data.Map.Strict as Map
import qualified Data.Set as Set
import qualified Data.Text as Text
import qualified Data.Text.Encoding as Encoding
import Json (J(..), encodeJson, field, object)
import Sha256 (sha256)

data Span = Span
  { spanStart :: Integer, spanEnd :: Integer
  , spanStartLine :: Integer, spanStartCol :: Integer
  , spanEndLine :: Integer, spanEndCol :: Integer
  } deriving (Eq, Show)

data Kind = Leaf | AllOf | AnyOf | Unsupported String deriving (Eq, Show)

data Requirement = Requirement
  { reqKind :: Kind, reqId :: String, reqPath :: [String], reqSpan :: Span
  , reqMembers :: [Requirement], reqPointer :: String
  } deriving (Eq, Show)

data Provision = Provision
  { provId :: String, provPath :: [String], provSpan :: Span
  , provRequirements :: [Requirement], provChildren :: [Provision]
  , provDefinitions :: Bool, provPointer :: String
  } deriving (Eq, Show)

data Diagnostic = Diagnostic
  { diagCode :: String, diagStage :: String, diagSeverity :: String, diagPath :: String
  , diagSpan :: Maybe Span, diagParameters :: [(String, String)]
  } deriving (Eq, Show)

data Input = Evaluate Provision (Map.Map String Bool) Integer
           | Validate Bool [Diagnostic]

data Trace = Trace
  { traceBranch :: String, traceId :: String, traceKind :: String
  , tracePath :: [String], traceSpan :: Span, traceValue :: Bool
  , traceChildren :: [String]
  }

data Branch = Branch
  { branchId :: String, branchPath :: [String], branchValue :: Bool
  , branchTraceIds :: [String]
  }

protocol, resultSchema, fragmentName :: String
protocol = "yuho.kernel-protocol/v1"
resultSchema = "yuho.kernel-result/v1"
fragmentName = "ClosedBooleanBranches-v1"

asString :: J -> Either String String
asString (JStr value) = Right value
asString _ = Left "expected string"

asBool :: J -> Either String Bool
asBool (JBool value) = Right value
asBool _ = Left "expected Boolean"

asInteger :: J -> Either String Integer
asInteger (JNum value) = Right value
asInteger _ = Left "expected integer"

asArray :: J -> Either String [J]
asArray (JArr value) = Right value
asArray _ = Left "expected array"

asObject :: J -> Either String [(String, J)]
asObject (JObj value) = Right value
asObject _ = Left "expected object"

named :: (J -> Either String a) -> String -> J -> Either String a
named decoder key value = field key value >>= decoder

strings :: J -> Either String [String]
strings value = asArray value >>= traverse asString

decodeSpan :: J -> Either String Span
decodeSpan value = do
  start <- named asInteger "start" value
  end <- named asInteger "end" value
  startLine <- named asInteger "start_line" value
  startCol <- named asInteger "start_col" value
  endLine <- named asInteger "end_line" value
  endCol <- named asInteger "end_col" value
  if start < 0 || end < start || startLine < 1 || endLine < startLine
     || startCol < 1 || endCol < 1
    then Left "invalid span"
    else Right (Span start end startLine startCol endLine endCol)

decodeRequirement :: String -> J -> Either String Requirement
decodeRequirement pointer value = do
  kindText <- named asString "kind" value
  identifier <- named asString "id" value
  path <- field "path" value >>= strings
  sourceSpan <- field "span" value >>= decodeSpan
  let kind = case kindText of
        "leaf" -> Leaf
        "all" -> AllOf
        "any" -> AnyOf
        _ -> Unsupported kindText
  members <- case kind of
    Leaf -> Right []
    Unsupported _ -> Right []
    _ -> do
      items <- field "members" value >>= asArray
      traverse (\(index, item) ->
        decodeRequirement (pointer ++ "/members/" ++ show index) item) (zip [0 :: Int ..] items)
  pure (Requirement kind identifier path sourceSpan members pointer)

decodeProvision :: String -> J -> Either String Provision
decodeProvision pointer value = do
  identifier <- named asString "id" value
  path <- field "path" value >>= strings
  sourceSpan <- field "span" value >>= decodeSpan
  definitions <- named asBool "definitions" value
  reqValues <- field "requirements" value >>= asArray
  requirements <- traverse (\(index, item) ->
    decodeRequirement (pointer ++ "/requirements/" ++ show index) item)
    (zip [0 :: Int ..] reqValues)
  childValues <- field "children" value >>= asArray
  children <- traverse (\(index, item) ->
    decodeProvision (pointer ++ "/children/" ++ show index) item)
    (zip [0 :: Int ..] childValues)
  pure (Provision identifier path sourceSpan requirements children definitions pointer)

decodeDiagnostic :: J -> Either String Diagnostic
decodeDiagnostic value = do
  code <- named asString "code" value
  stage <- named asString "stage" value
  severity <- named asString "severity" value
  pointer <- named asString "path" value
  sourceSpan <- field "span" value >>= \item -> case item of
    JNull -> Right Nothing
    _ -> Just <$> decodeSpan item
  parameterValue <- field "parameters" value >>= asObject
  parameters <- traverse (\(key, item) -> (,) key <$> asString item) parameterValue
  if severity /= "error" && severity /= "warning" && severity /= "info"
    then Left "invalid diagnostic severity"
    else Right (Diagnostic code stage severity pointer sourceSpan parameters)

decodeInput :: J -> Either String Input
decodeInput root = do
  schema <- named asString "input_schema" root
  fragment <- named asString "fragment" root
  if schema /= "yuho.kernel-input/v1" || fragment /= fragmentName
    then Left "unsupported input schema or fragment"
    else pure ()
  source <- field "source" root
  sourceText <- named asString "text" source
  sourceHash <- named asString "sha256" source
  _ <- named asString "path" source
  if sha256 (Encoding.encodeUtf8 (Text.pack sourceText)) /= sourceHash
    then Left "source SHA-256 mismatch"
    else pure ()
  policy <- field "policy" root
  _ <- named asString "reference_date" policy
  maxNodes <- named asInteger "max_nodes" policy
  if maxNodes < 1 || maxNodes > 1024 then Left "invalid max_nodes" else pure ()
  operation <- named asString "operation" root
  case operation of
    "evaluate" -> do
      program <- field "program" root >>= decodeProvision "/program"
      factValues <- field "facts" root >>= asObject
      facts <- traverse (\(key, item) -> (,) key <$> asBool item) factValues
      pure (Evaluate program (Map.fromList facts) maxNodes)
    "validate" -> do
      parserResult <- field "parser_result" root
      accepted <- named asBool "accepted" parserResult
      _ <- named asString "source_version" parserResult
      diagnosticValues <- field "diagnostics" parserResult >>= asArray
      diagnostics <- traverse decodeDiagnostic diagnosticValues
      pure (Validate accepted diagnostics)
    _ -> Left "unsupported operation"

spanJson :: Span -> J
spanJson value = object
  [ ("start", JNum (spanStart value)), ("end", JNum (spanEnd value))
  , ("start_line", JNum (spanStartLine value)), ("start_col", JNum (spanStartCol value))
  , ("end_line", JNum (spanEndLine value)), ("end_col", JNum (spanEndCol value))
  ]

diagnosticJson :: Diagnostic -> J
diagnosticJson value = object
  [ ("code", JStr (diagCode value)), ("stage", JStr (diagStage value))
  , ("severity", JStr (diagSeverity value)), ("path", JStr (diagPath value))
  , ("span", maybe JNull spanJson (diagSpan value))
  , ("parameters", object [(key, JStr item) | (key, item) <- diagParameters value])
  ]

makeDiagnostic :: String -> String -> String -> Maybe Span -> [(String, String)] -> Diagnostic
makeDiagnostic code stage pointer sourceSpan parameters =
  Diagnostic code stage "error" pointer sourceSpan parameters

validateProgram :: Integer -> Map.Map String Bool -> Provision -> Either Diagnostic ()
validateProgram maxNodes facts root = do
  let nodes = allNodes root
  if toInteger (length nodes) > maxNodes
    then Left (makeDiagnostic "KINV004" "validate" "/program" (Just (provSpan root)) [])
    else pure ()
  _ <- foldM checkNode Set.empty nodes
  let leaves = [(reqId item, reqSpan item) | Right item <- nodes, reqKind item == Leaf]
  mapM_ (\(identifier, sourceSpan) ->
    if Map.member identifier facts then Right ()
    else Left (makeDiagnostic "KINV001" "validate" ("/facts/" ++ identifier)
               (Just sourceSpan) [("id", identifier)])) leaves
  where
    checkNode seen node = case node of
      Left provision ->
        if Set.member (provId provision) seen
          then Left (makeDiagnostic "KINV002" "validate"
                     (provPointer provision ++ "/id") (Just (provSpan provision))
                     [("id", provId provision)])
          else Right (Set.insert (provId provision) seen)
      Right requirement -> case reqKind requirement of
        Unsupported kind ->
          Left (makeDiagnostic "KCAP001" "capability"
                (reqPointer requirement ++ "/kind") (Just (reqSpan requirement))
                [("kind", kind)])
        _ | Set.member (reqId requirement) seen ->
              Left (makeDiagnostic "KINV002" "validate"
                    (reqPointer requirement ++ "/id") (Just (reqSpan requirement))
                    [("id", reqId requirement)])
          | reqKind requirement /= Leaf && null (reqMembers requirement) ->
              Left (makeDiagnostic "KINV004" "validate"
                    (reqPointer requirement ++ "/members") (Just (reqSpan requirement)) [])
          | otherwise -> Right (Set.insert (reqId requirement) seen)

allNodes :: Provision -> [Either Provision Requirement]
allNodes provision =
  Left provision : concatMap requirementNodes (provRequirements provision)
  ++ concatMap allNodes (provChildren provision)

requirementNodes :: Requirement -> [Either Provision Requirement]
requirementNodes value = Right value : concatMap requirementNodes (reqMembers value)

leafBranches :: Provision -> [Requirement] -> [(Provision, [Requirement])]
leafBranches provision inherited =
  let direct = inherited ++ provRequirements provision
      descendants = concatMap (\child -> leafBranches child direct) (provChildren provision)
  in if not (null descendants) then descendants
     else if null direct then [] else [(provision, direct)]

evaluateRequirement :: Map.Map String Bool -> String -> Requirement
                    -> Either Diagnostic (Bool, [Trace])
evaluateRequirement facts currentBranchId requirement =
  case reqKind requirement of
    Unsupported kind ->
      Left (makeDiagnostic "KCAP001" "capability" (reqPointer requirement ++ "/kind")
            (Just (reqSpan requirement)) [("kind", kind)])
    Leaf -> case Map.lookup (reqId requirement) facts of
      Nothing -> Left (makeDiagnostic "KINV001" "validate" ("/facts/" ++ reqId requirement)
                       (Just (reqSpan requirement)) [("id", reqId requirement)])
      Just value -> Right (value, [makeTrace value []])
    kind -> do
      evaluated <- traverse (evaluateRequirement facts currentBranchId) (reqMembers requirement)
      let values = map fst evaluated
          value = if kind == AllOf then and values else or values
          children = concatMap snd evaluated
      Right (value, makeTrace value (map reqId (reqMembers requirement)) : children)
  where
    makeTrace value children =
      Trace currentBranchId (reqId requirement) (kindName (reqKind requirement))
        (reqPath requirement) (reqSpan requirement) value children

kindName :: Kind -> String
kindName Leaf = "leaf"
kindName AllOf = "all"
kindName AnyOf = "any"
kindName (Unsupported name) = name

evaluateBranch :: Map.Map String Bool -> (Provision, [Requirement])
               -> Either Diagnostic (Branch, [Trace])
evaluateBranch facts (provision, requirements) = do
  evaluated <- traverse (evaluateRequirement facts (provId provision)) requirements
  let value = and (map fst evaluated)
      trace = concatMap snd evaluated
  Right (Branch (provId provision) (provPath provision) value (map traceId trace), trace)

boolName :: Bool -> String
boolName True = "true"
boolName False = "false"

branchJson :: Branch -> J
branchJson branch = object
  [ ("id", JStr (branchId branch)), ("path", JArr (map JStr (branchPath branch)))
  , ("status", JStr (boolName (branchValue branch)))
  , ("trace_ids", JArr (map JStr (branchTraceIds branch)))
  ]

traceJson :: Trace -> J
traceJson trace = object
  [ ("branch", JStr (traceBranch trace)), ("id", JStr (traceId trace))
  , ("kind", JStr (traceKind trace)), ("path", JArr (map JStr (tracePath trace)))
  , ("span", spanJson (traceSpan trace)), ("value", JStr (boolName (traceValue trace)))
  , ("children", JArr (map JStr (traceChildren trace)))
  ]

innerDigest :: J -> String
innerDigest root =
  let operation = named asString "operation" root
      keys = case operation of
        Right "validate" -> ["input_schema","fragment","source","parser_result","policy"]
        _ -> ["input_schema","fragment","source","program","facts","policy"]
      pairs = traverse (\key -> (,) key <$> field key root) keys
  in case pairs of
    Left _ -> sha256 BS.empty
    Right values -> sha256 (Encoding.encodeUtf8 (Text.pack (encodeJson (object values))))

baseResult :: String -> String -> String -> String -> [Branch] -> [Trace] -> [Diagnostic] -> J
baseResult requestId digestValue status kind branches traces diagnostics = object
  [ ("protocol", JStr protocol), ("request_id", JStr requestId)
  , ("result_schema", JStr resultSchema), ("fragment", JStr fragmentName)
  , ("input_digest", JStr digestValue), ("status", JStr status)
  , ("provision_kind", JStr kind), ("branches", JArr (map branchJson branches))
  , ("trace", JArr (map traceJson traces))
  , ("diagnostics", JArr (map diagnosticJson diagnostics))
  ]

run :: J -> J
run root =
  let requestId = either (const "?") id (named asString "request_id" root)
      digestValue = innerDigest root
      reject = baseResult requestId digestValue "rejected" "none" [] []
      received = either (const "") id (named asString "protocol" root)
  in if received /= protocol
     then reject [makeDiagnostic "KPROT001" "protocol" "/protocol" Nothing
                   [("received", received)]]
     else case decodeInput root of
       Left message -> reject [makeDiagnostic "KDEC001" "decode" "/" Nothing
                                [("reason", message)]]
       Right (Validate accepted diagnostics) ->
         baseResult requestId digestValue (if accepted then "true" else "rejected")
           "none" [] [] diagnostics
       Right (Evaluate program facts maxNodes) ->
         case validateProgram maxNodes facts program of
           Left diagnostic -> reject [diagnostic]
           Right () ->
             case traverse (evaluateBranch facts) (leafBranches program []) of
               Left diagnostic -> reject [diagnostic]
               Right evaluated ->
                 let branches = map fst evaluated
                     traces = concatMap snd evaluated
                     kind = if null branches then "definition_only" else "executable"
                 in baseResult requestId digestValue (boolName (any branchValue branches))
                      kind branches traces []

decodeFailure :: String -> J
decodeFailure message =
  baseResult "?" (sha256 BS.empty) "rejected" "none" [] []
    [makeDiagnostic "KDEC001" "decode" "/" Nothing [("reason", message)]]

File: experiments/language-spike/ocaml/src/kernel.ml
open Json

let ( let* ) = Result.bind

type span = {
  start : int;
  end_ : int;
  start_line : int;
  start_col : int;
  end_line : int;
  end_col : int;
}

type kind = Leaf | All | Any | Unsupported of string

type requirement = {
  kind : kind;
  id : string;
  path : string list;
  span : span;
  members : requirement list;
  pointer : string;
}

type provision = {
  id : string;
  path : string list;
  span : span;
  requirements : requirement list;
  children : provision list;
  definitions : bool;
  pointer : string;
}

type diagnostic = {
  code : string;
  stage : string;
  severity : string;
  path : string;
  span : span option;
  parameters : (string * string) list;
}

type input =
  | Evaluate of provision * (string * bool) list * int
  | Validate of bool * diagnostic list

type trace = {
  branch : string;
  id : string;
  kind : string;
  path : string list;
  span : span;
  value : bool;
  children : string list;
}

type branch = {
  id : string;
  path : string list;
  value : bool;
  trace_ids : string list;
}

let protocol = "yuho.kernel-protocol/v1"
let result_schema = "yuho.kernel-result/v1"
let fragment = "ClosedBooleanBranches-v1"

let get_string = function Str value -> Ok value | _ -> Error "expected string"
let get_bool = function Bool value -> Ok value | _ -> Error "expected Boolean"
let get_int = function Num value -> Ok value | _ -> Error "expected integer"
let get_list = function Arr value -> Ok value | _ -> Error "expected array"
let get_object = function Obj value -> Ok value | _ -> Error "expected object"

let named decoder key value =
  let* item = field key value in
  decoder item

let rec collect = function
  | [] -> Ok []
  | result :: rest ->
      let* value = result in
      let* tail = collect rest in
      Ok (value :: tail)

let strings value =
  let* values = get_list value in
  collect (List.map get_string values)

let decode_span value =
  let* start = named get_int "start" value in
  let* end_ = named get_int "end" value in
  let* start_line = named get_int "start_line" value in
  let* start_col = named get_int "start_col" value in
  let* end_line = named get_int "end_line" value in
  let* end_col = named get_int "end_col" value in
  if start < 0 || end_ < start || start_line < 1 || end_line < start_line
     || start_col < 1 || end_col < 1
  then Error "invalid span"
  else Ok { start; end_; start_line; start_col; end_line; end_col }

let rec decode_requirement pointer value =
  let* kind_text = named get_string "kind" value in
  let* id = named get_string "id" value in
  let* path_value = field "path" value in
  let* path = strings path_value in
  let* span_value = field "span" value in
  let* span = decode_span span_value in
  let kind =
    match kind_text with
    | "leaf" -> Leaf
    | "all" -> All
    | "any" -> Any
    | other -> Unsupported other
  in
  let* members =
    match kind with
    | Leaf | Unsupported _ -> Ok []
    | All | Any ->
        let* member_value = field "members" value in
        let* values = get_list member_value in
        values
        |> List.mapi (fun index child ->
             decode_requirement (pointer ^ "/members/" ^ string_of_int index) child)
        |> collect
  in
  Ok { kind; id; path; span; members; pointer }

let rec decode_provision pointer value =
  let* id = named get_string "id" value in
  let* path_value = field "path" value in
  let* path = strings path_value in
  let* span_value = field "span" value in
  let* span = decode_span span_value in
  let* definitions = named get_bool "definitions" value in
  let* req_value = field "requirements" value in
  let* req_values = get_list req_value in
  let* requirements =
    req_values
    |> List.mapi (fun index child ->
         decode_requirement (pointer ^ "/requirements/" ^ string_of_int index) child)
    |> collect
  in
  let* child_value = field "children" value in
  let* child_values = get_list child_value in
  let* children =
    child_values
    |> List.mapi (fun index child ->
         decode_provision (pointer ^ "/children/" ^ string_of_int index) child)
    |> collect
  in
  Ok { id; path; span; requirements; children; definitions; pointer }

let decode_diagnostic value =
  let* code = named get_string "code" value in
  let* stage = named get_string "stage" value in
  let* severity = named get_string "severity" value in
  let* path = named get_string "path" value in
  let* span_value = field "span" value in
  let* span =
    match span_value with Null -> Ok None | other ->
      let* decoded = decode_span other in
      Ok (Some decoded)
  in
  let* parameter_value = field "parameters" value in
  let* parameter_pairs = get_object parameter_value in
  let* parameters =
    parameter_pairs
    |> List.map (fun (key, item) ->
         let* text = get_string item in
         Ok (key, text))
    |> collect
  in
  if not (List.mem severity [ "error"; "warning"; "info" ])
  then Error "invalid diagnostic severity"
  else Ok { code; stage; severity; path; span; parameters }

let decode_input root =
  let* schema = named get_string "input_schema" root in
  let* fragment_name = named get_string "fragment" root in
  if schema <> "yuho.kernel-input/v1" || fragment_name <> fragment
  then Error "unsupported input schema or fragment"
  else
    let* source = field "source" root in
    let* source_text = named get_string "text" source in
    let* source_hash = named get_string "sha256" source in
    let* _source_path = named get_string "path" source in
    if Sha256.sha256 source_text <> source_hash then Error "source SHA-256 mismatch"
    else
      let* policy = field "policy" root in
      let* _date = named get_string "reference_date" policy in
      let* max_nodes = named get_int "max_nodes" policy in
      if max_nodes < 1 || max_nodes > 1024 then Error "invalid max_nodes"
      else
        let* operation = named get_string "operation" root in
        match operation with
        | "evaluate" ->
            let* program_value = field "program" root in
            let* program = decode_provision "/program" program_value in
            let* fact_value = field "facts" root in
            let* fact_pairs = get_object fact_value in
            let* facts =
              fact_pairs
              |> List.map (fun (key, item) ->
                   let* value = get_bool item in
                   Ok (key, value))
              |> collect
            in
            Ok (Evaluate (program, facts, max_nodes))
        | "validate" ->
            let* parser_result = field "parser_result" root in
            let* accepted = named get_bool "accepted" parser_result in
            let* _source_version = named get_string "source_version" parser_result in
            let* diagnostic_value = field "diagnostics" parser_result in
            let* diagnostic_values = get_list diagnostic_value in
            let* diagnostics = collect (List.map decode_diagnostic diagnostic_values) in
            Ok (Validate (accepted, diagnostics))
        | _ -> Error "unsupported operation"

let span_json (value : span) =
  Obj [
    "start", Num value.start; "end", Num value.end_;
    "start_line", Num value.start_line; "start_col", Num value.start_col;
    "end_line", Num value.end_line; "end_col", Num value.end_col;
  ]

let diagnostic_json (value : diagnostic) =
  Obj [
    "code", Str value.code; "stage", Str value.stage;
    "severity", Str value.severity; "path", Str value.path;
    "span", (match value.span with None -> Null | Some location -> span_json location);
    "parameters", Obj (List.map (fun (key, text) -> key, Str text) value.parameters);
  ]

let make_diagnostic code stage path span parameters =
  { code; stage; severity = "error"; path; span; parameters }

type node = Provision_node of provision | Requirement_node of requirement

let rec requirement_nodes (value : requirement) =
  Requirement_node value :: List.concat_map requirement_nodes value.members

let rec all_nodes (value : provision) =
  Provision_node value
  :: (List.concat_map requirement_nodes value.requirements
      @ List.concat_map all_nodes value.children)

module Id_set = Set.Make (String)

let validate_program max_nodes facts root =
  let nodes = all_nodes root in
  if List.length nodes > max_nodes then
    Error (make_diagnostic "KINV004" "validate" "/program" (Some root.span) [])
  else
    let check_node seen = function
      | Provision_node item ->
          if Id_set.mem item.id seen then
            Error (make_diagnostic "KINV002" "validate"
                     (item.pointer ^ "/id") (Some item.span) [ "id", item.id ])
          else Ok (Id_set.add item.id seen)
      | Requirement_node item ->
          (match item.kind with
           | Unsupported name ->
               Error (make_diagnostic "KCAP001" "capability"
                        (item.pointer ^ "/kind") (Some item.span) [ "kind", name ])
           | _ when Id_set.mem item.id seen ->
               Error (make_diagnostic "KINV002" "validate"
                        (item.pointer ^ "/id") (Some item.span) [ "id", item.id ])
           | Leaf -> Ok (Id_set.add item.id seen)
           | All | Any when item.members = [] ->
               Error (make_diagnostic "KINV004" "validate"
                        (item.pointer ^ "/members") (Some item.span) [])
           | All | Any -> Ok (Id_set.add item.id seen))
    in
    let* _seen =
      List.fold_left
        (fun state node ->
          let* seen = state in
          check_node seen node)
        (Ok Id_set.empty) nodes
    in
    let leaves : requirement list =
      List.filter_map
        (function
          | Requirement_node ({ kind = Leaf; _ } as item) -> Some item
          | _ -> None)
        nodes
    in
    List.fold_left
      (fun state (item : requirement) ->
        let* () = state in
        if List.mem_assoc item.id facts then Ok ()
        else Error (make_diagnostic "KINV001" "validate"
                      ("/facts/" ^ item.id) (Some item.span) [ "id", item.id ]))
      (Ok ()) leaves

let rec leaf_branches (provision : provision) inherited =
  let direct = inherited @ provision.requirements in
  let descendants =
    List.concat_map (fun child -> leaf_branches child direct) provision.children
  in
  if descendants <> [] then descendants
  else if direct = [] then []
  else [ provision, direct ]

let kind_name = function
  | Leaf -> "leaf"
  | All -> "all"
  | Any -> "any"
  | Unsupported name -> name

let rec evaluate_requirement facts branch_id (item : requirement) =
  let make_trace value children =
    {
      branch = branch_id; id = item.id; kind = kind_name item.kind;
      path = item.path; span = item.span; value; children;
    }
  in
  match item.kind with
  | Unsupported name ->
      Error (make_diagnostic "KCAP001" "capability"
               (item.pointer ^ "/kind") (Some item.span) [ "kind", name ])
  | Leaf ->
      (match List.assoc_opt item.id facts with
       | None ->
           Error (make_diagnostic "KINV001" "validate"
                    ("/facts/" ^ item.id) (Some item.span) [ "id", item.id ])
       | Some value -> Ok (value, [ make_trace value [] ]))
  | All | Any ->
      let* evaluated =
        collect (List.map (evaluate_requirement facts branch_id) item.members)
      in
      let values = List.map fst evaluated in
      let value = if item.kind = All then List.for_all Fun.id values else List.exists Fun.id values in
      let children = List.concat_map snd evaluated in
      Ok (value, make_trace value (List.map (fun (member : requirement) -> member.id) item.members) :: children)

let evaluate_branch facts ((provision : provision), requirements) =
  let* evaluated =
    collect (List.map (evaluate_requirement facts provision.id) requirements)
  in
  let value = List.for_all (fun (result, _) -> result) evaluated in
  let trace = List.concat_map snd evaluated in
  Ok ({ id = provision.id; path = provision.path; value;
        trace_ids = List.map (fun (item : trace) -> item.id) trace }, trace)

let bool_name value = if value then "true" else "false"

let branch_json (value : branch) =
  Obj [
    "id", Str value.id; "path", Arr (List.map (fun item -> Str item) value.path);
    "status", Str (bool_name value.value);
    "trace_ids", Arr (List.map (fun item -> Str item) value.trace_ids);
  ]

let trace_json (value : trace) =
  Obj [
    "branch", Str value.branch; "id", Str value.id;
    "kind", Str value.kind; "path", Arr (List.map (fun item -> Str item) value.path);
    "span", span_json value.span; "value", Str (bool_name value.value);
    "children", Arr (List.map (fun item -> Str item) value.children);
  ]

let inner_digest root =
  let operation = named get_string "operation" root in
  let keys =
    match operation with
    | Ok "validate" ->
        [ "input_schema"; "fragment"; "source"; "parser_result"; "policy" ]
    | _ ->
        [ "input_schema"; "fragment"; "source"; "program"; "facts"; "policy" ]
  in
  let pairs =
    keys
    |> List.map (fun key ->
         let* value = field key root in
         Ok (key, value))
    |> collect
  in
  match pairs with
  | Error _ -> Sha256.sha256 ""
  | Ok items -> Sha256.sha256 (Json.encode (Obj items))

let base_result request_id digest status kind branches traces diagnostics =
  Obj [
    "protocol", Str protocol;
    "request_id", Str request_id;
    "result_schema", Str result_schema;
    "fragment", Str fragment;
    "input_digest", Str digest;
    "status", Str status;
    "provision_kind", Str kind;
    "branches", Arr (List.map branch_json branches);
    "trace", Arr (List.map trace_json traces);
    "diagnostics", Arr (List.map diagnostic_json diagnostics);
  ]

let run root =
  let request_id =
    match named get_string "request_id" root with Ok value -> value | Error _ -> "?"
  in
  let digest = inner_digest root in
  let reject diagnostic =
    base_result request_id digest "rejected" "none" [] [] [ diagnostic ]
  in
  let received =
    match named get_string "protocol" root with Ok value -> value | Error _ -> ""
  in
  if received <> protocol then
    reject (make_diagnostic "KPROT001" "protocol" "/protocol" None
              [ "received", received ])
  else
    match decode_input root with
    | Error message ->
        reject (make_diagnostic "KDEC001" "decode" "/" None
                  [ "reason", message ])
    | Ok (Validate (accepted, diagnostics)) ->
        base_result request_id digest
          (if accepted then "true" else "rejected") "none" [] [] diagnostics
    | Ok (Evaluate (program, facts, max_nodes)) ->
        (match validate_program max_nodes facts program with
         | Error diagnostic -> reject diagnostic
         | Ok () ->
             (match collect (List.map (evaluate_branch facts) (leaf_branches program [])) with
              | Error diagnostic -> reject diagnostic
              | Ok evaluated ->
                  let branches = List.map fst evaluated in
                  let traces = List.concat_map snd evaluated in
                  let kind = if branches = [] then "definition_only" else "executable" in
                  base_result request_id digest
                    (bool_name (List.exists (fun (item : branch) -> item.value) branches))
                    kind branches traces []))

let decode_failure message =
  base_result "?" (Sha256.sha256 "") "rejected" "none" [] []
    [ make_diagnostic "KDEC001" "decode" "/" None [ "reason", message ] ]
